File: Scala/src/main/scala/uk/org/winder/maths/Fibonacci.scala
package uk.org.winder.maths

import scala.annotation.tailrec
import scala.math.sqrt

object Fibonacci {

  val zero = BigInt(0)
  val one = BigInt(1)
  val two = BigInt(2)

 private def validate(n:BigInt):Unit = {
    if (n < zero) throw new IllegalArgumentException("Parameter must be a non-negative integer.")
  }

  def iterativeWhile(n:Int):BigInt = iterativeWhile(BigInt(n))
  def iterativeWhile(n:BigInt):BigInt = {
    validate(n)
    var i = n
    var result = zero
    var next = one
    while (i > zero) {
      val temporary = result
      result = next
      next = temporary + next
      i = i - one
    }
    result
  }

  def iterativeFor(n:Int):BigInt = iterativeFor(BigInt(n))
  def iterativeFor(n:BigInt):BigInt = {
    validate(n)
    var result = zero
    var next = one
    for (i <- zero until n) {
      val temporary = result
      result = next
      next = temporary + next
    }
    result
  }

  def naïveRecursive(n:Int):BigInt = naïveRecursive(BigInt(n))
  def naïveRecursive(n:BigInt):BigInt = {
    validate(n)
    if (n < two) n
    else naïveRecursive(n - one) + naïveRecursive(n - two)
  }

  def tailRecursive(n:Int):BigInt = tailRecursive(BigInt(n))
  def tailRecursive(n:BigInt):BigInt = {
    validate(n)
    @tailrec def iteration(n:BigInt, next:BigInt = one, result:BigInt = zero):BigInt = n match {
      case `zero` => result
      case _ => iteration(n - one, next + result, next)
    }
    iteration(n)
   }

  private val memo = scala.collection.mutable.Map(BigInt(0) -> BigInt(0), BigInt(1) -> BigInt(1))
  def memoizedRecursive(n:Int):BigInt = memoizedRecursive(BigInt(n))
  def memoizedRecursive(n:BigInt):BigInt = {
    validate(n)
    if (! memo.contains(n)) {
      memo += n -> (memoizedRecursive(n - one) + memoizedRecursive(n - two))
    }
    memo(n)
  }

  def foldLeftive(n:Int):BigInt = foldLeftive(BigInt(n))
  def foldLeftive(n:BigInt):BigInt = {
    validate(n)
    if (n < two) n
    else ((zero, one) /: (one until n))((a, _) => (a._2, a._1 + a._2))._2
  }

  def zipive(n:Int):BigInt = zipive(BigInt(n))
  def zipive(n:BigInt):BigInt = {
    validate(n)
    assert(n < Int.MaxValue)
    lazy val fib: Stream[BigInt] = zero #:: one #:: fib.zip(fib.tail).map{pair => pair._1 + pair._2}
    fib.drop(n.toInt).head
  }

  def closedForm(n:Int):BigInt = closedForm(BigInt(n))
  def closedForm(n:BigInt):BigInt = {
    validate(n)
    val sqrt5 = BigDecimal(sqrt(5))
    return (((1 + sqrt5) / 2).pow(n.intValue) / sqrt5 + 0.5).toBigInt
  }

}

File: Java/src/main/java/uk/org/winder/maths/Fibonacci.java
package uk.org.winder.maths;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import static java.lang.Math.sqrt;

public class Fibonacci {
    private static final BigInteger zero = BigInteger.ZERO;
    private static final BigInteger one = BigInteger.ONE;
    private static final BigInteger two = BigInteger.TWO;

    private static void validate(final BigInteger n) {
        if (n.compareTo(zero) < 0) {
            throw new IllegalArgumentException("index must be positive.");
        }
    }

    public static BigInteger iterative(final Long n) {
        return iterative(BigInteger.valueOf(n));
    }

    public static BigInteger iterative(final BigInteger n) {
        validate(n);
        var result = zero;
        var next = one;
        for (var i = zero; i.compareTo(n) < 0; i = i.add(one)) {
            var temporary = result;
            result = next;
            next = temporary.add(next);
        }
        return result;
    }

    public static BigInteger naiveRecursive(final Long n) {
        return naiveRecursive(BigInteger.valueOf(n));
    }

    public static BigInteger naiveRecursive(final BigInteger n) {
        validate(n);
        return
                n.compareTo(zero) == 0 ? zero :
                        n.compareTo(one) == 0 ? one :
                                naiveRecursive(n.subtract(one)).add(naiveRecursive(n.subtract(two)));
    }

    private static Map<BigInteger, BigInteger> memo = new HashMap<>();
    static {
        memo.put(zero, zero);
        memo.put(one, one);
    }

    public static BigInteger memoizedRecursive(final Long n) {
        return memoizedRecursive(BigInteger.valueOf(n));
    }

    public static BigInteger memoizedRecursive(final BigInteger n) {
        validate(n);
        if (!memo.containsKey(n)) {
            memo.put(n, memoizedRecursive(n.subtract(one)).add(memoizedRecursive(n.subtract(two))));
        }
        return memo.get(n);
    }

    public static BigInteger closedForm(final Long n) {
        return closedForm(BigInteger.valueOf(n));
    }

    public static BigInteger closedForm(final BigInteger n) {
        validate(n);
        //  Use what is known as Binet's Formula. cf. https://en.wikipedia.org/wiki/Fibonacci_number#Binet's_formula
        final var one = BigDecimal.ONE;
        final var two = BigDecimal.valueOf(2);
        final var half = BigDecimal.valueOf(0.5);
        final var sqrt5 = BigDecimal.valueOf(sqrt(5));
        final var n_i = n.intValue();
        final var phi = sqrt5.add(one).divide(two).pow(n_i);
        final var psi = sqrt5.subtract(one).divide(two).pow(n_i);
        return phi.subtract(psi).divide(sqrt5, RoundingMode.HALF_UP).add(half).toBigInteger();
    }

    public static class Sequence {
        private final static ArrayList<BigInteger> numbers = new ArrayList<>();
        static {
            numbers.add(zero);
            numbers.add(one);
        }

        public static class Iterator implements java.util.Iterator<BigInteger> {
            private int currentIndex = 0;

            @Override
            public boolean hasNext() { return true; }

            @Override
            public BigInteger next() { return Sequence.getAt(currentIndex++); }

            @Override
            public void remove() { throw new UnsupportedOperationException(); }
        }

        public static Iterator iterator() { return new Iterator(); }

        public static BigInteger getAt(final int index) {
            if (index < 0) {
                throw new IllegalArgumentException("Negative index not allowed.");
            }
            if (index >= numbers.size()) {
                numbers.ensureCapacity(index + 1);
                for (var i = numbers.size(); i <= index; ++i) {
                    numbers.add(numbers.get(i - 1).add(numbers.get(i - 2)));
                }
            }
            return numbers.get(index);
        }
    }
}
